Update vec_f64_ppc.h to use POWER9 test data class instructions.

munroesj52 opened the pull request:
Fixes after POWER9 testing in mambo.

	* src/vec_f64_ppc.h (vec_all_isinff64 [_ARCH_PWR9]):
	Correct DMCX const value for vec_test_data_class().
        Should be 0x30 for +- Infinity.
	(vec_all_issubnormalf64 [!_ARCH_PWR9]): Correct logic for
	POWER8 vec_all_issubnormalf64 () code. Should be ALL (exponent < 1)
        AND ALL (NOT 0.0).
	(vec_any_isinff64 [_ARCH_PWR9]): Use vec_any_eq() for compare. Using
        vec_all_eq was a pasto. Should use vec_any_eq to convert the
        test_data_class to vector CC or int bool.
	(vec_any_isnanf64 [_ARCH_PWR9]): Ditto.
	(vec_any_isnormalf64 [_ARCH_PWR9]): Ditto.
	(vec_any_issubnormalf64 [_ARCH_PWR9]): Ditto.
	(vec_any_iszerof64 [_ARCH_PWR9]): Ditto.

	* src/testsuite/arith128_test_f64.c (test_double_all_is):
	Add test case for vec_all_issubnormalf64 for 2x
	__DBL_DENORM_MIN__ values. Correct test to detect false (was testing
        true) for vec_all_issubnormalf64 ({0.0, __DBL_DENORM_MIN__}) case.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

ThinkOpenly commented:
For your commit message, you have "Dido".  I presume you meant "Ditto"?  Or perhaps "Likewise"?

munroesj52 commented:
Yes I can't spell and yes I meant Ditto.

None of these are obvious. 

In this case the failures did not show until tested compiled for -mcpu=power9 and tested ON POWER9 (or mambo -s power9). Also there are two cases under _ARCH_PWR9 depending on the compiler version (AT10 vs AT11/AT12).

It took me 2 weeks to build my own AT12 for Ubuntu 14.04 (not supported), which would install on Debian Jessie (also not supported) which will run on mambo).

So there where several pasto's where the _ARCH_PWR9 code for vec_any_isinf/nan/ etc where copied from the vec_f32_ppc.h vec_all_is* code. The result was that vec_any classification code worked when compiled for -mcpu=power8 but failed for -mcpu=power9. Because I used a vec_all_eq to convert the vector predicate result from vec_test_data_class() into a condition code or integer bool. But using vec_all_eq should have been vec_any_eq. 

In the case of vec_all_issubnormalf64 the power8 logic was incorrect but the test case was also wrong and hide the problem. This became obvious when I fixed the power9 logic. So fixed the test case (should be false for vec_all_issubnormalf64 and true for vec_any_issubnormalf64). Also added a new test (2 x __DBL_DENORM_MIN__) which should be true for both vec_all_issubnormalf64 and vec_any_issubnormalf64). And then I corrected the logical for power8 vec_all_issubnormalf64.

Then I tested the whole mess again. The cross product of {AT10, AT11, AT12) x (P7, P8, P9)

ThinkOpenly commented:
can you add the substantive contents of your comment above to the commit message?

OK for commit then.
